Backend/src/routes/api.js

```diff
@@ -157,11 +171,100 @@ ROUTER.get("/process-recorded-data", (req, res) => {
 
 function recordData(data) {
   fs.appendFile(RECORDED_DATA_PATH + currentSession + ".bin", Buffer.concat([data, Buffer.alloc(1, true)]),
-                (err) => {
-                  if(err) {
-                    console.error("ERROR: Error while appending to file");
-                  }
-  });
+    (err) => {
+      if (err) {
+        console.error("ERROR: Error while appending to file");
+      }
+    });
+}
+
+
+
+// --------------------------------------------------------------------------------------------------------------------
+// Performance improvements
+// --------------------------------------------------------------------------------------------------------------------
+
+ROUTER.post(CONSTANTS.ROUTES.UPDATE_GRAPHS_METADATA, (req, res) => {
+  // console.log("(BACKEND)needed-graph-metadata:", req.body.meta)
+  if (req.body) {
+    updateGraphsMetadata(req.body)
+    res.send({ status: "SUCCESS" }).status(200)
+  } else {
+    res.send({ status: "EMPTY-REQ" }).status(200)
+  }
+
+});
+
+function getSingleValues(jsonData) {
+  let newJson = {}
+
+  for (const key in jsonData) {
+    if (jsonData.hasOwnProperty(key)) {
+      newJson[key] = [jsonData[key][0]]
+    }
+  }
+
+  return newJson
+}
+
+function filterGraphsToSend() {
+  let obj = {};
+  let maxNumValues = 1; // Needs to be 1 so that we have at least one timestamp
+
+  graphsToSend.map((element) => {
+    const numValues = (element[1] / 600) * X_AXIS_CAP; // TODO Update this with a non-static X_AXIS_CAP
+
+    obj[`${element[0]}`] = frontendData[`${element[0]}`].slice(0, Math.round(numValues));
+
+    if(numValues > maxNumValues) {
+      maxNumValues = numValues;
+    }
+  })
+
+  obj["timestamps"] = frontendData["timestamps"].slice(0, Math.round(maxNumValues));
+  /* Logs for checking lengths of each dataset and timestamps
+  console.log("Max num values:", Math.round(maxNumValues))
+  for (const key in obj) {
+    console.log(`${key}: ${obj[key].length}`);
+  }*/
+
+  return obj
+}
+
+function updateGraphsMetadata(data) {
+  graphsToSend = []
+
+  for (const [key, value] of Object.entries(data)) {
+    graphsMetadata[`${key}`] = value
+  }
+
+  for (const key in graphsMetadata) {
+    const data = graphsMetadata[key];
+    updateGraphsToSend(data)
+  }
+
+}
+
+function updateGraphsToSend(data) {
+  console.log("[----------------NEEDED GRAPHS REQUEST (updateGraphsToSend)-----------------]")
+  if (data) {
+    console.log(":----Looking into Item: ")
+    for (let i = 0; i < data?.datasets?.length; i++) {
+      const dataset = data?.datasets[i];
+      const historyLength = data?.historyLength;
+      const datasetIdx = graphsToSend.map(element => element[0]).indexOf(dataset);
+
+      if (datasetIdx === -1) {
+        graphsToSend.push([dataset, historyLength]);
```

Comment:
A comment or two might be good here to make it more clear that it's searching for duplicates, it took me a second to figure out what it was doing